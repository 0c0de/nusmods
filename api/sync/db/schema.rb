# This file is auto-generated from the current state of the database. Instead
# of editing this file, please use the migrations feature of Active Record to
# incrementally modify your database, and then regenerate this schema definition.
#
# Note that this schema.rb definition is the authoritative source for your
# database schema. If you need to create the application database on another
# system, you should be using db:schema:load, not running all the migrations
# from scratch. The latter is a flawed and unsustainable approach (the more migrations
# you'll amass, the slower it'll run and the greater likelihood for issues).
#
# It's strongly recommended that you check this file into your version control system.

ActiveRecord::Schema.define(version: 2018_08_17_014544) do

  create_table "acad_years", force: :cascade do |t|
    t.string "name", null: false
    t.integer "school_id", null: false
    t.datetime "created_at", null: false
    t.datetime "updated_at", null: false
    t.index ["school_id"], name: "index_acad_years_on_school_id"
  end

  create_table "courses", force: :cascade do |t|
    t.string "code", null: false
    t.text "title", null: false
    t.text "description", null: false
    t.string "slug", null: false
    t.integer "credit", null: false
    t.string "department"
    t.string "workload"
    t.string "preclusion"
    t.string "prerequisite"
    t.string "corequisite"
    t.integer "semester_id", null: false
    t.datetime "created_at", null: false
    t.datetime "updated_at", null: false
    t.index ["semester_id"], name: "index_courses_on_semester_id"
  end

  create_table "faculties", force: :cascade do |t|
    t.string "name", null: false
    t.string "slug", null: false
    t.integer "school_id", null: false
    t.datetime "created_at", null: false
    t.datetime "updated_at", null: false
    t.index ["school_id"], name: "index_faculties_on_school_id"
  end

  create_table "lessons", force: :cascade do |t|
    t.string "code", null: false
    t.string "class_type", null: false
    t.string "day", null: false
    t.string "week", null: false
    t.datetime "start_at", null: false
    t.datetime "end_at", null: false
    t.integer "course_id", null: false
    t.integer "venue_id"
    t.datetime "created_at", null: false
    t.datetime "updated_at", null: false
    t.index ["course_id"], name: "index_lessons_on_course_id"
    t.index ["venue_id"], name: "index_lessons_on_venue_id"
  end

  create_table "schools", force: :cascade do |t|
    t.string "short_name", null: false
    t.string "long_name", null: false
    t.string "slug", null: false
    t.datetime "created_at", null: false
    t.datetime "updated_at", null: false
  end

  create_table "semesters", force: :cascade do |t|
    t.string "name", null: false
    t.datetime "start_at", null: false
    t.datetime "end_at", null: false
    t.integer "acad_year_id", null: false
    t.datetime "created_at", null: false
    t.datetime "updated_at", null: false
    t.index ["acad_year_id"], name: "index_semesters_on_acad_year_id"
  end

  create_table "venues", force: :cascade do |t|
    t.string "code", null: false
    t.string "name", null: false
    t.string "slug", null: false
    t.string "floor"
    t.decimal "lat", precision: 10, scale: 7
    t.decimal "long", precision: 10, scale: 7
    t.decimal "altitude", precision: 10, scale: 7
    t.integer "faculty_id"
    t.datetime "created_at", null: false
    t.datetime "updated_at", null: false
    t.index ["faculty_id"], name: "index_venues_on_faculty_id"
  end

end
