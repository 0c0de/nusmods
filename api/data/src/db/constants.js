/**
 * Lookups from tables (camelCase) to tables (snake_case)
 */
const TABLES = {
  schools: 'schools',
  terms: 'terms',
  departments: 'departments',
  venues: 'venues',
  courses: 'courses',
  lessons: 'lessons',
};

/**
 * Lookups from columns (camelCase) to columns (snake_case)
 */
const COLUMNS = {
  altitude: 'altitude',
  code: 'code',
  corequisite: 'corequisite',
  courseId: 'course_id',
  createdAt: 'created_at',
  day: 'day',
  departmentId: 'department_id',
  description: 'description',
  endsAt: 'ends_at',
  floor: 'floor',
  id: 'id',
  lat: 'lat',
  lessonId: 'lesson_id',
  lng: 'lng',
  longName: 'long_name',
  name: 'name',
  preclusion: 'preclusion',
  prerequisite: 'prerequisite',
  schoolId: 'school_id',
  shortName: 'short_name',
  startsAt: 'starts_at',
  termId: 'term_id',
  title: 'title',
  type: 'type',
  updatedAt: 'updated_at',
  value: 'value',
  venueId: 'venue_id',
  week: 'week',
  workload: 'workload',
};

/**
 * Lookups from enum (camelCase) to enum (snake_case)
 */
const ENUMS = {
  days: 'days_enum',
};

/**
 * Lookups from table (camelCase) to foreign key (snake_case)
 */
const TABLE_TO_FOREIGN_KEYS = {
  schools: COLUMNS.schoolId,
  terms: COLUMNS.termId,
  departments: COLUMNS.departmentId,
  venues: COLUMNS.venueId,
  courses: COLUMNS.courseId,
  lessons: COLUMNS.lessonId,
};

const DAYS = ['Monday', 'Tuesday', 'Wednesday', 'Thursday', 'Friday', 'Saturday', 'Sunday'];

module.exports = {
  TABLES,
  COLUMNS,
  ENUMS,
  TABLE_TO_FOREIGN_KEYS,
  DAYS,
};
