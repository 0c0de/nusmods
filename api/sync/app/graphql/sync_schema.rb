# frozen_string_literal: true

class SyncSchema < GraphQL::Schema
  mutation(Types::MutationType)
  query(Types::QueryType)
end
