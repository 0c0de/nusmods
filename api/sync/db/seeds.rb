# This file should contain all the record creation needed to seed the database with its default values.
# The data can then be loaded with the rails db:seed command (or created alongside the database with db:setup).
#
# Examples:
#
#   movies = Movie.create([{ name: 'Star Wars' }, { name: 'Lord of the Rings' }])
#   Character.create(name: 'Luke', movie: movies.first)

@num_fakes = 5

def new_acad_years
  Array.new(@num_fakes) do
    AcadYear.new(
      name: Faker::Cannabis.strain
    )
  end
end

@num_fakes.times do
  long_name = Faker::University.name
  short_name = long_name.split.map(&:first).join

  School.create(
    short_name: short_name,
    long_name: long_name,
    slug: short_name.downcase,
    acad_years: new_acad_years
  )
end
