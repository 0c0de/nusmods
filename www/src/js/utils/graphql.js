import { ApolloLink } from 'apollo-link';
import { ApolloClient } from 'apollo-client';
import { HttpLink } from 'apollo-link-http';
import { InMemoryCache } from 'apollo-cache-inmemory';
import ActionCable from 'actioncable';
import ActionCableLink from 'graphql-ruby-client/subscriptions/ActionCableLink';

const cable = ActionCable.createConsumer('ws://localhost:4000/cable');

const httpLink = new HttpLink({
  uri: 'http://localhost:4000/graphql',
  credentials: 'same-origin',
});

const hasSubscriptionOperation = ({ query: { definitions } }) =>
  definitions.some(
    ({ kind, operation }) => kind === 'OperationDefinition' && operation === 'subscription',
  );

const link = ApolloLink.split(hasSubscriptionOperation, new ActionCableLink({ cable }), httpLink);

const client = new ApolloClient({
  link,
  cache: new InMemoryCache(),
});

export default client;
