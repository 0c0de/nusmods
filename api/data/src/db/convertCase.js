import { isObjectLike, invert } from 'lodash';
import { TABLES, COLUMNS } from './constants';

/**
 * Only do conversion according to known column and table names in
 * constants.js because knex's own migration tables and queries also
 * run though these functions.
 */

const CAMEL_CASE_TO_SNAKE_CASE = { ...TABLES, ...COLUMNS };
const SNAKE_CASE_TO_CAMEL_CASE = invert(CAMEL_CASE_TO_SNAKE_CASE);

function snakeCase(camelCaseStr) {
  return CAMEL_CASE_TO_SNAKE_CASE[camelCaseStr] || camelCaseStr;
}
function camelCase(snakeCaseStr) {
  return SNAKE_CASE_TO_CAMEL_CASE[snakeCaseStr] || snakeCaseStr;
}

function camelCaseResult(x) {
  if (Array.isArray(x)) {
    return x.map(camelCaseResult);
  } else if (isObjectLike(x)) {
    const obj = {};
    Object.entries(x).forEach(([key, value]) => {
      obj[camelCase(key)] = value;
    });
    return obj;
  }
  return camelCase(x);
}

export default {
  snakeCase,
  camelCase: camelCaseResult,
};
