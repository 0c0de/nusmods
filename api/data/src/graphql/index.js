import { gql } from 'apollo-server';
import _ from 'lodash';
import { TABLES } from '../db/constants';
import db from '../db/database';
import { DateScalarType } from './customScalars';

const typeDefs = gql`
  scalar Date

  type School {
    id: ID!
    createdAt: Date!
    updatedAt: Date!
    longName: String!
    shortName: String
  }

  type Term {
    id: ID!
    createdAt: Date!
    updatedAt: Date!
    name: String!
    startsAt: Date!
    endsAt: Date!
  }

  # Describes a module, may span different semesters
  type Course {
    id: ID!
    createdAt: Date!
    updatedAt: Date!
    code: String!
    title: String!
    description: String
    value: Float
    workload: String
    prerequisite: String
    corequisite: String
  }

  # Describes a particular module for a semester
  type CourseInfo {
    semester: Int
    examDate: String
    examOpenBook: Boolean
    examDuration: String
    examVenue: String
    timetable: [Lesson]
  }

  # Bidding stats for Cors
  type CorsBiddingStats {
    quota: Int
    bidders: Int
    lowestBid: Int
    lowestSuccessfulBid: Int
    highestBid: Int
    faculty: String
    studentAcctType: String
    acadYear: String
    semester: Int
    round: String
    group: String
  }

  # A lesson conducted, may it be a lecture, laboratory or lecture
  type Lesson {
    classNo: String!
    lessonType: String!
    weekText: String!
    dayText: String!
    startTime: String!
    endTime: String!
    venue: String!
  }

  # the schema allows the following query:
  type Query {
    schools(first: Int, offset: Int): [School!]!
    terms(schoolId: ID!, first: Int, offset: Int): [Term!]!
    courses(termId: ID!, first: Int, offset: Int): [Course!]!
    course(termId: ID!, code: String!): Course
  }

  type Mutation {
    term(schoolId: ID!, name: String!, startsAt: Date!, endsAt: Date!): ID!
  }

  schema {
    query: Query
    mutation: Mutation
  }
`;

const resolvers = {
  Query: {
    schools(root, { first, offset }) {
      return db
        .table(TABLES.schools)
        .select()
        .limit(first || -1)
        .offset(offset);
    },
    terms(root, { schoolId, first, offset }) {
      return db
        .table(TABLES.terms)
        .select()
        .where({ schoolId })
        .limit(first || -1)
        .offset(offset);
    },
    courses(root, { termId, first, offset }) {
      return db
        .table(TABLES.courses)
        .select()
        .where({ termId })
        .limit(first || -1)
        .offset(offset);
    },
    course(root, { termId, code }) {
      return db
        .table(TABLES.courses)
        .select()
        .where({ termId, code })
        .first();
    },
  },
  Mutation: {
    term(root, { schoolId, name, startsAt, endsAt }) {
      return db
        .table(TABLES.terms)
        .insert({ schoolId, name, startsAt, endsAt })
        .then((x) => x[0]);
    },
  },
  Date: DateScalarType,
};

export default {
  typeDefs,
  resolvers,
};
