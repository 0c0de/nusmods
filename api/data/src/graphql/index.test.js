import { graphql } from 'graphql';
import { makeExecutableSchema } from 'apollo-server';
import index from './';
import db from '../db/database';

const schema = makeExecutableSchema(index);

const gql = (x) => x.raw[0]; // identify function for template literals

describe('queries', () => {
  beforeAll(async () => {
    await db.migrate.rollback();
    await db.migrate.latest();
    await db.seed.run();
  });

  afterAll(async () => {
    await db.migrate.rollback();
  });

  describe('course queries', () => {

    it('should be empty when courses are not found', async () => {
      const query = gql`
        query {
          courses(termId: -1) {
            code
          }
        }
      `;
      const { data } = await graphql(schema, query);

      expect(data).not.toBeNull();
      expect(data.courses).toEqual([]);
    });

    it('should be not be null when courses are found', async () => {
      const query = gql`
        query {
          courses(termId: 1) {
            code
          }
        }
      `;
      const { data } = await graphql(schema, query);

      expect(data).not.toBeNull();
      expect(data).toMatchSnapshot();
    });

    it('should return everything when first and offset are not specified', async () => {
      const query = gql`
        query {
          courses(termId: 1) {
            code
          }
        }
      `;
      const {
        data: { courses },
      } = await graphql(schema, query);

      expect(courses).toHaveLength(2);
    });

    it('should return first n elements when specified', async () => {
      const query = gql`
        query {
          courses(termId: 1, first: 1) {
            code
          }
        }
      `;
      const {
        data: { courses },
      } = await graphql(schema, query);

      expect(courses).toHaveLength(1);
    });

    it('should return offset n elements when specified', async () => {
      const query = gql`
        query {
          courses(termId: 1, offset: 2) {
            code
          }
        }
      `;
      const {
        data: { courses },
      } = await graphql(schema, query);

      expect(courses).toHaveLength(0);
    });

    it('should return first n and offset n elements when specified', async () => {
      const query = gql`
        query {
          courses(termId: 1, first: 1, offset: 1) {
            id
          }
        }
      `;
      const {
        data: { courses },
      } = await graphql(schema, query);

      expect(courses).toHaveLength(1);
      expect(courses[0].id).toBe('2');
    });

    it('should be null when course is not found', async () => {
      const query = gql`
        query {
          course(termId: 2, code: "CS2100") {
            code
          }
        }
      `;
      const {
        data: { course },
      } = await graphql(schema, query);

      expect(course).toBeNull();
    });

    it('should not be null when course is valid', async () => {
      const query = gql`
        query {
          course(termId: 1, code: "CS2100") {
            code
            title
            credit
            department
            description
          }
        }
      `;
      const { data } = await graphql(schema, query);

      expect(data).not.toBeNull();
      expect(data).toMatchSnapshot();
    });
  });

  describe('terms query', () => {
    it('should return terms when school id is valid', async () => {
      const query = gql`
        query {
          terms(schoolId: 1) {
            id
          }
        }
      `;
      const {
        data: { terms },
      } = await graphql(schema, query);
      expect(terms).toHaveLength(1);
    });
  });
});
