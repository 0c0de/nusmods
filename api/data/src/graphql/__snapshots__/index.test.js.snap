// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`queries course queries should be not be null when courses are found 1`] = `
Object {
  "courses": Array [
    Object {
      "code": "CS1101S",
    },
    Object {
      "code": "CS2100",
    },
  ],
}
`;

exports[`queries course queries should not be null when course is valid 1`] = `undefined`;
